client/state/lib/middleware.js

```diff
@@ -22,14 +36,45 @@ const updateSelectedSiteForAnalytics = ( dispatch, action, getState ) => {
 	analytics.setSiteCount( siteCount );
 };
 
-export const handlers = {
-	[ ANALYTICS_SUPER_PROPS_UPDATE ]: updateSelectedSiteForAnalytics,
+/**
+ * Sets the selectedSiteId for lib/cart/store
+ *
+ * @param {function} dispatch - redux dispatch function
+ * @param {object}   action   - the dispatched action
+ * @param {function} getState - redux getState function
+ */
+const updateSelectedSiteForCart = ( dispatch, action, getState ) => {
+	const state = getState();
+	const selectedSiteId = getSelectedSiteId( state );
+	cartStore.setSelectedSiteId( selectedSiteId );
 };
 
-export const libraryMiddleware = ( { dispatch, getState } ) => ( next ) => ( action ) => {
-	if ( handlers.hasOwnProperty( action.type ) ) {
-		handlers[ action.type ]( dispatch, action, getState );
+const handler = ( dispatch, action, getState ) => {
+	switch ( action.type ) {
+		case ANALYTICS_SUPER_PROPS_UPDATE:
+			return updateSelectedSiteForAnalytics( dispatch, action, getState );
+
+		case SELECTED_SITE_SET:
+		case SITE_RECEIVE:
+		case SITES_RECEIVE:
+		case SITES_UPDATE:
+			// Wait a tick for the reducer to update the state tree
+			setTimeout( () => {
+				updateSelectedSiteForCart( dispatch, action, getState );
+			}, 0 );
+			return;
+
+		case ROUTE_SET:
+			// Special case, no site actions are dispatched at all for no-site
+			// checkout flow
+			if ( action.path === '/checkout/no-site' ) {
```

Comment:
Actually, I think we have a cleaner approach if we add a route handler to reset selected site state. h/t @ockham It should be up soon